File: 5/main.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <fstream>

using namespace std;

void DFS1(int v, vector<vector<int>>& graph, vector<bool>& visited, stack<int>& finishStack) {
    visited[v] = true;
    for (int i = 0; i < graph[v].size(); i++) {
        if (!visited[i] && graph[v][i] > 0) {
            DFS1(i, graph, visited, finishStack);
        }
    }
    finishStack.push(v);
}

void DFS2(int v, vector<vector<int>>& transposeGraph, vector<bool>& visited, vector<int>& component) {
    visited[v] = true;
    component.push_back(v);
    for (int i = 0; i < transposeGraph[v].size(); i++) {
        if (!visited[i] && transposeGraph[v][i] > 0) {
            DFS2(i, transposeGraph, visited, component);
        }
    }
}

int main() {
    ifstream fin("input.txt");
    ofstream fout("output.txt");

    int n;
    fin >> n; // Чтение количества вершин графа

    vector<vector<int>> graph(n, vector<int>(n));
    vector<vector<int>> transposeGraph(n, vector<int>(n));
    vector<bool> visited(n, false);
    stack<int> finishStack;
    vector<vector<int>> components;

    // Чтение матрицы смежности из файла
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            fin >> graph[i][j];
            transposeGraph[j][i] = graph[i][j]; // Транспонирование графа
        }
    }
    fin.close();

    // 1-й проход DFS для определения порядка завершения вершин
    for (int i = 0; i < n; ++i) {
        if (!visited[i]) {
            DFS1(i, graph, visited, finishStack);
        }
    }

    // Очистка флагов посещения для второго прохода
    fill(visited.begin(), visited.end(), false);

    // 2-й проход DFS на транспонированном графе
    while (!finishStack.empty()) {
        int v = finishStack.top();
        finishStack.pop();
        if (!visited[v]) {
            vector<int> component;
            DFS2(v, transposeGraph, visited, component);
            components.push_back(component);
        }
    }

    // Вывод результатов
    fout << "Количество сильно связных компонент: " << components.size() << endl;
    for (const auto& component : components) {
        for (int v : component) {
            fout << v << " ";
        }
        fout << endl;
    }

    fout.close();
    return 0;
}

File: 5/main.py
# Функция для чтения матрицы смежности из файла
def read_adjacency_matrix(filename):
    with open(filename, 'r') as file:
        size = int(file.readline().strip())
        matrix = [[int(num) for num in line.split()] for line in file]
    return matrix

# Функция для записи результата в файл
def write_result(filename, result):
    with open(filename, 'w') as file:
        file.write(result)

# Обход в глубину (DFS) для графа
def dfs(graph, node, visited, stack):
    visited[node] = True  # Помечаем узел как посещенный
    for i in range(len(graph[node])):
        if graph[node][i] and not visited[i]:  # Для каждой смежной вершины
            dfs(graph, i, visited, stack)  # Рекурсивно вызываем DFS
    stack.append(node)  # Добавляем узел в стек после посещения всех смежных вершин

# Функция для транспонирования графа
def transpose_graph(graph):
    # Транспонирование матрицы смежности
    return [list(row) for row in zip(*graph)]

# Алгоритм Косарайю для нахождения сильно связных компонент
def kosaraju(graph):
    visited = [False] * len(graph)  # Инициализируем список посещенных вершин
    stack = []  # Стек для хранения порядка завершения вершин
    
    # 1-й проход DFS для заполнения стека
    for i in range(len(graph)):
        if not visited[i]:
            dfs(graph, i, visited, stack)
    
    # Транспонируем граф
    transposed_graph = transpose_graph(graph)
    visited = [False] * len(graph)  # Сбрасываем список посещенных вершин
    components = []  # Список для хранения сильно связных компонент
    
    # 2-й проход DFS на транспонированном графе
    while stack:
        node = stack.pop()  # Берем вершину из стека
        if not visited[node]:
            component = []
            dfs(transposed_graph, node, visited, component)  # Выполняем DFS для вершины
            components.append(component)  # Добавляем сильно связную компоненту в список
    
    return components


input_filename = 'input.txt'
output_filename = 'output.txt'

adjacency_matrix = read_adjacency_matrix(input_filename)
print(transpose_graph(adjacency_matrix))
components = kosaraju(adjacency_matrix)


result = f"Количество сильно связных компонент: {len(components)}\nСостав компонент:\n"
i = 0
for component in components:
    result += f"Компонент {i+1}: {component}\n"
    i += 1
write_result(output_filename, result)